provider/google: Support for unmanaged instance groups (google_compute_instance_group)

ajcrowe opened the pull request:
I've added support for creating unmanaged Google Instance Groups. I've not been able to complete the suite of acceptance tests but I think they should work.

I've used a couple of functions found in `google_compute_target_pool.go`

* `convertInstances()` https://github.com/hashicorp/terraform/blob/master/builtin/providers/google/resource_compute_target_pool.go#L103
* `convertStringArr()` https://github.com/hashicorp/terraform/blob/master/builtin/providers/google/resource_compute_target_pool.go#L79
* `calcAddRemove()` https://github.com/hashicorp/terraform/blob/master/builtin/providers/google/resource_compute_target_pool.go#L167

Is having a dependency on these ok?

Let me know if anything needs changing more than happy to amend.

sparkprime commented on builtin/providers/google/resource_compute_instance_group.go:
```diff
@@ -0,0 +1,301 @@
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+
+			"network": &schema.Schema{
```
According to the API docs, this is not computed

ajcrowe replied:
This is strange, the library docs list it as `[Output Only]` see https://godoc.org/google.golang.org/api/compute/v1#InstanceGroup

From my testing the adding of instances will populate the network into the state. or leave it blank should no instance be defined when it's created.

If you later add instance the network is populated correctly. However the network remains even after all instances are removed, this could be something to change?

lwander commented on builtin/providers/google/resource_compute_instance_group.go:
```diff
@@ -0,0 +1,301 @@
+				ForceNew: true,
+			},
+
+			"fingerprint": &schema.Schema{
```
This isn't necessary to have in the schema. In `update`, you want to do a 3-way merge between the old TF state, the new TF state, and the current server state. When you retrieve that server state during the update you get the fingerprint you need for the update to succeed. Since this fingerprint is only valid and needed for that single update, it doesn't make sense to keep in the schema. I know that some resources like `google_compute_instance` keep fingerprint in the schema, but it's a mistake and a pain to remove.

ajcrowe replied:
I've the removed `fingerprint` as suggested it was unnecessary. 

In terms of the 3 way merge would you suggest I just call `resourceComputeInstanceGroupRead` before to make sure no other changes have occurred? Or can I assume the state to be correct?

lwander replied:
Well, the merge is really only interesting for the `instances` list, which your read doesn't update. What the other GCE resources do in a similar case is in the `Read` (which is automatically called before `Update` by Terraform) they only record changes to instances that were initially added to the local TF file. For example, you specify instances `A, B, C`, someone else adds `D` and deletes `A`, and then `Read` records `B, C` for the local state. The rational is, if you recorded `D` as existing in the local state, then the local state would show an attribute existing that wasn't declared in the local TF file, which would then be deleted in the next `Update`. Recording the deletion of `A` is important, as it keeps the user state in sync with the server state. The last part of the merge is then between the new and old state in `Update`.

ajcrowe replied:
Ok I understand, so you would want the manually added `D` to persist and Terraform to only manage the instances that have been declared?

lwander replied:
Yes that's exactly right.

On Wed, Dec 2, 2015 at 4:06 AM, Alex Crowe <notifications@github.com> wrote:

> In builtin/providers/google/resource_compute_instance_group.go
> <https://github.com/hashicorp/terraform/pull/4087#discussion_r46390373>:
>
> > +		Delete: resourceComputeInstanceGroupDelete,
> > +
> > +		Schema: map[string]*schema.Schema{
> > +			"name": &schema.Schema{
> > +				Type:     schema.TypeString,
> > +				Required: true,
> > +				ForceNew: true,
> > +			},
> > +
> > +			"description": &schema.Schema{
> > +				Type:     schema.TypeString,
> > +				Optional: true,
> > +				ForceNew: true,
> > +			},
> > +
> > +			"fingerprint": &schema.Schema{
>
> Ok I understand, so you would want the manually added D to persist and
> Terraform to only manage the instances that have been declared?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hashicorp/terraform/pull/4087/files#r46390373>.
>

lwander commented on builtin/providers/google/resource_compute_instance_group.go:
```diff
@@ -0,0 +1,301 @@
+			"instances": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: false,
```
Did you mean for this to be `ForceNew`? You seem to check for updates in `update`.

ajcrowe replied:
This is set to false, so it's unnecessary and I've removed it

lwander replied:
Ah, silly me

lwander commented:
@ajcrowe Provided tests all pass? LGTM, could you squash before I merge?

ajcrowe commented:
I've squashed the commit and rebased to master. I've also removed the `getNamedPorts` function as this  was pull into an update to the `google_compute_instance_group_manager` resource. 

I've updated the docs to reflect the need to always use full self_link urls for instances and added a simple check for this. If you need any other changes let me know

lwander commented:
Sweet! One last tiny nitpick - could you reformat your test configs to take a name parameter like [here](https://github.com/hashicorp/terraform/blob/master/builtin/providers/google/resource_compute_instance_test.go#L499), and pass in a name with a random suffix like [so](https://github.com/hashicorp/terraform/blob/master/builtin/providers/google/resource_compute_instance_test.go#L39)? The motivation is explained in #4514. Your current chosen names shouldn't cause any conflicts, but we'd like to adopt this practice for all tests just in case.

ajcrowe commented:
I've updated the acctests, I haven't been able to run them myself so please let me know if they are broken.

lwander commented:
The tests are broken, 1, you're missing quotes around the instances resource you are referencing with `${}`, and two, those instances need to exist for the test to pass.

ajcrowe commented:
This should be sorted now, sorry for the oversight!

ajcrowe commented:
Thanks @lwander, sorry I missed this! I've pull your change in.

lwander commented:
All set! sorry for the long wait.

ajcrowe commented:
No worries Lars, pleased I could contribute to a great tool.
